lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+  EulerAngles angles;
+  /// Updates to the global phase
+  double phase;
+
+  /// This logic is based on https://arxiv.org/pdf/quant-ph/9503016 and its
+  /// corresponding explanation in https://threeplusone.com/pubs/on_gates.pdf,
+  /// Section 4.
+  void decompose() {
+    using namespace std::complex_literals;
+
+    /// Rescale the input unitary matrix, `u`, to be special unitary.
+    /// Extract a phase factor, `phase`, so that `det(inv_phase * unitary) = 1`
+    auto det = (matrix[0] * matrix[3]) - (matrix[1] * matrix[2]);
+    phase = 0.5 * std::atan2(det.imag(), det.real());
+    auto inv_phase = std::exp(-1i * phase);
+
+    std::array<std::complex<double>, 4> special_unitary;
+    std::transform(matrix.begin(), matrix.end(), special_unitary.begin(),
+                   [&](auto element) { return element * inv_phase; });
+
+    auto abs_00 = std::abs(special_unitary[0]);
+    auto abs_01 = std::abs(special_unitary[1]);
+
+    if (abs_00 >= abs_01)
+      angles.beta = 2 * std::acos(abs_00);
+    else
+      angles.beta = 2 * std::asin(abs_01);
+
+    auto half_sum_alpha_gamma =
+        std::atan2(special_unitary[3].imag(), special_unitary[3].real());
+    auto half_diff_alpha_gamma =
+        std::atan2(special_unitary[2].imag(), special_unitary[2].real());
+
+    angles.alpha = half_sum_alpha_gamma + half_diff_alpha_gamma;
+    angles.gamma = half_sum_alpha_gamma - half_diff_alpha_gamma;
+  }
+
+  BasisZYZ(std::vector<std::complex<double>> vec) {
+    std::move(vec.begin(), vec.end(), matrix.begin());
+  }
+};
+
+class CustomUnitaryPattern
+    : public OpRewritePattern<quake::CustomUnitarySymbolOp> {
+
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::CustomUnitarySymbolOp customOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto parentModule = customOp->getParentOfType<ModuleOp>();
+
+    Location loc = customOp->getLoc();
+    auto targets = customOp.getTargets();
+    auto controls = customOp.getControls();
+
+    /// Get the global constant holding the concrete matrix corresponding to
+    /// this custom operation invocation
+    StringRef generatorName = customOp.getGenerator().getRootReference();
+    auto globalOp =
+        parentModule.lookupSymbol<cudaq::cc::GlobalOp>(generatorName);
+
+    /// The decomposed sequence of quantum operations are in a function
+    auto pair = generatorName.split(".rodata");
+    std::string funcName = pair.first.str() + ".kernel" + pair.second.str();
+
+    /// If the replacement function doesn't exist, create it here
+    if (!parentModule.lookupSymbol<LLVM::LLVMFuncOp>(funcName)) {
+
+      auto unitary = cudaq::readGlobalConstantArray(globalOp);
+
+      /// TODO: Expand the logic to decompose upto 4-qubit operations
+      if (unitary.size() != 4)
+        return customOp.emitError(
+            "Decomposition of only single qubit custom operations supported.");
```

Comment:
I'm thinking the pass raising a fatal error here might be too Draconian in general. Do we really want the compiler to crash if the matrix isn't 2x2?

Reply:
The hardware targets cannot process the `quake.custom_op` unless it is decomposed, thus cannot proceed further.

Reply:
Sure. But there are other places, like target codegen, that can make that decision, or potentially make some other transformation.

Additionally, this pass is written as a greedy pass, which connotates that it is a best effort pass. That is, it is not an all-or-nothing showstopper (like the codegen pipeline necessarily is).

Reply:
Okay, updated in a0630832cfc57b43b7744e9ff50e96c233e8d742, sot his pass only returns 'failure' (with warning) and then the error message comes from codegen.